### PR TITLE
Further Development of Union Overlap Experiment

### PR DESCRIPTION
**Pull request by rmccall84**
Experiment runs UnionPooler on input from a Temporal Memory with and
without training. Compute the overlap between the Union SDR representations in these two
conditions over time.

### CODE REVIEW

**Comment by rmccall84** [comment]
@scottpurdy I've got a new baseline of the overlap based on the discussion we had on Friday. I'm happy with the results, which I'll post shortly to a wiki page. This PR is ready to be merged, please give it your review. Thank you.

File: union_pooling/tests/unit/union_pooler_test.py [hunk 1]
```diff
@@ -51,6 +51,19 @@ def testDecayPoolingActivation(self):
...
 
+  def testDecayPoolingActivationSpecifiedDecayRate(self):
+    self.unionPooler._decayFunctionSlope = 10
+    self.unionPooler._poolingActivation = numpy.array([0, 10, 20, 30, 40],
+                                                      dtype=REAL_DTYPE)
+    expected = numpy.array([0, 0, 10, 20, 30])
+
+    result = self.unionPooler._decayPoolingActivation()
+
+    self.assertEquals(len(expected), len(result))
```
Comment (scottpurdy) [inline union_pooling/tests/unit/union_pooler_test.py]: Can you use `self.assertSequenceEqual` to replace this plus the loop?

File: union_pooling/union_pooling/experiments/union_sdr_overlap/plot_experiment.py [hunk 2]
```diff
@@ -24,18 +24,36 @@
...
+    os.makedirs(imgOutputPath)
+
   print "Computing Union SDR overlap between SDR traces in following dir:"
   print inputPath + "\n"
 
   files = os.listdir(inputPath)
-  assert len(files) == 2
+  if len(files) != 2:
+    print "Found {0} files at input path. Requires exactly 2.".format(inputPath)
+    return
```
Comment (scottpurdy) [inline union_pooling/union_pooling/experiments/union_sdr_overlap/plot_experiment.py]: Call `sys.exit` with a non-zero exit code to indicate a failure.

File: union_pooling/union_pooling/experiments/union_sdr_overlap/union_sdr_overlap_experiment.py [hunk 3]
```diff
@@ -97,12 +89,44 @@ def writeDefaultMetrics(outputDir, experiment, patternDimensionality,
 
 
 
+def writeMetricTrace(experiment, traceName, outputDir, outputFileName):
+  """
+  Assume trace elements can be converted to list.
+  :param experiment:
```
Comment (scottpurdy) [inline union_pooling/union_pooling/experiments/union_sdr_overlap/union_sdr_overlap_experiment.py]: Give these descriptions?

File: union_pooling/union_pooling/experiments/union_sdr_overlap/union_sdr_overlap_experiment.py [hunk 4]
```diff
@@ -51,26 +54,15 @@
...
-    outputFile.flush()
+def getBurstingColumnsStats(experiment):
+  traceData = experiment.tm.mmGetTraceUnpredictedActiveColumns().data
+  resetData = experiment.tm.mmGetTraceResets().data
+  countTrace = [0 if resetData[x] else len(traceData[x])
+                for x in xrange(len(traceData))]
+  mean = numpy.mean(countTrace)
+  stdDev = numpy.std(countTrace)
+  maximum = max(countTrace)
+  return [mean, stdDev, maximum]
```
Comment (scottpurdy) [inline union_pooling/union_pooling/experiments/union_sdr_overlap/union_sdr_overlap_experiment.py]: Probably better to return a tuple (immutable)

**Comment by scottpurdy** [comment]
Done with review. :+1:

File: union_pooling/tests/unit/union_pooler_test.py (hunk 1 above)
Comment (rmccall84) [inline union_pooling/tests/unit/union_pooler_test.py]: Good point. 

I get a ValueError when I used this assert function. The issue is that the arrays are numpy and comparing numpy arrays with == is ambiguous (http://stackoverflow.com/questions/10062954/valueerror-the-truth-value-of-an-array-with-more-than-one-element-is-ambiguous). 

I found the following works: self.assertTrue(numpy.array_equal(expected, result))

File: union_pooling/union_pooling/experiments/union_sdr_overlap/union_sdr_overlap_experiment.py (hunk 4 above)
Comment (rmccall84) [inline union_pooling/union_pooling/experiments/union_sdr_overlap/union_sdr_overlap_experiment.py]: nice to know, thanks.

File: union_pooling/tests/unit/union_pooler_test.py (hunk 1 above)
Comment (scottpurdy) [inline union_pooling/tests/unit/union_pooler_test.py]: Yeah that looks good

**Pull request merged**